### ISSUE TITLE
.detach() behaves differently for dense tensors vs sparse tensors

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#34375]
## 🐛 Bug

When we call tensor.detach(), the operations supported on the detached tensor differ in case of sparse/dense.

## To Reproduce

```
import torch

t = torch.rand(3,3, requires_grad=True)
t1 = t.detach()
t1.zero_()

i = torch.LongTensor([[0, 1, 1],
                          [2, 0, 2]])
v = torch.FloatTensor([3, 4, 5])
sp = torch.sparse.FloatTensor(i, v, torch.Size([2,3]))
sp1 = sp.detach()
with torch.no_grad():
  sp1.zero_()
```

The last line `sp1.zero_()` throws the following exception:
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-74456cf884fa> in <module>
     11 sp1 = sp.detach()
     12 with torch.no_grad():
---> 13     sp1.zero_()

RuntimeError: resize_and_clear_ is not allowed on a Tensor created from .data or .detach().
If your intent is to change the metadata of a Tensor (such as sizes / strides / storage / storage_offset)
without autograd tracking the change, remove the .data / .detach() call and wrap the change in a `with torch.no_grad():` block.
For example, change:
    x.data.set_(y)
to:
    with torch.no_grad():
        x.set_(y)

```
## Expected behavior

We should have consistent behavior across sparse and dense tensors after detaching.

## Additional context

This was discovered as part of https://github.com/pytorch/pytorch/pull/33427.

cc @ezyang @albanD

**Comment by albanD** [comment]
There are consistent in the sense that if you try to resize_() a detached dense Tensor, you will get the same issue.
It just happens that `.zero_()` needs to resize the tensor if it is sparse.

What is your idea on how we can fix that?

**Comment by ezyang** [comment]
My claim is that we should start allowing metadata modification on `detach()`ed tensors. The semantics is that detach gives you a fresh metadata, and we've already had a release cycle where we forbid those modifications, and so now we can just change the semantics.

**Comment by albanD** [comment]
But what does it mean to do `a.detach().resize_(...)`?

**Comment by ezyang** [comment]
My suggestion is that it causes a resize of the `a.detach()` tensor, but doesn't resize `a`.

**Comment by albanD** [comment]
If you do that, then `my_sparse_tensor.detach().zero_()` will not zero out the original sparse Tensor. Which is not really the expected behavior...

**Comment by gchanan** [comment]
it's an interesting issue, because we put in that metadata error so people wouldn't do `a.detach().resize_(...)` and expect `a` to resize.  But this is really about how you look at detach -- is it a logical "view" of the Tensor (where you would expect `a.detach().zero_()` to zero out `a` or not.  We could also do some in-between thing where like...the implementation wipes out indices/values if it can detect that it is safe to do, otherwise it zeros out the values tensor.

We could also, separately, get rid of the metadata error for dense tensors.

**Comment by ezyang** [comment]
Yeah, this sucks. I can't think of any way around it besides gchanan's idea, but I think it's not a good idea to have a big pile of zeros hanging around either. Maybe we should just add a `nograd_zero_` method just for this case, blegh.

**Comment by albanD** [comment]
@gchanan  I think [our doc](https://pytorch.org/docs/stable/autograd.html#torch.Tensor.detach) is very clear about this:
- Returns a new Tensor: so changing metadata of the Tensor won't affect the original one
- Shares storage: so it is a view and changing the values will change the original Tensor.

The idea behind preventing metadata change was to avoid a silent BC-breaking change.
I would be in favor of Ed's idea:
> we've already had a release cycle where we forbid those modifications, and so now we can just change the semantics.
